File: apps/mobile/src/lib/nutrition/logic/rhythm-logic.ts
/**
 * Pure daily-rhythm helpers vendored from web `components/nutrition/sections/
 * daily-rhythm.tsx` (keep in sync). `COMPOSITION_COLORS` swaps the web's
 * `var(--nham-macro-*)` for resolved mobile tokens.
 */
import type { MacroPattern } from '@/lib/nutrition/types';
import { colors } from '~/theme/tokens';

export const KCAL_PER_GRAM = {
  protein: 4,
  carbohydrate: 4,
  fat: 9,
} as const;

export const COMPOSITION_KEYS = ['protein', 'carbohydrate', 'fat'] as const;
export type CompositionKey = (typeof COMPOSITION_KEYS)[number];

export const COMPOSITION_COLORS: Record<CompositionKey, string> = {
  protein: colors.macroProtein,
  carbohydrate: colors.macroCarbs,
  fat: colors.macroFat,
};

export const COMPOSITION_SHORT: Record<CompositionKey, string> = {
  protein: 'P',
  carbohydrate: 'C',
  fat: 'F',
};

export const ORDERED_MACROS = [
  'protein',
  'carbohydrate',
  'fat',
  'fiber',
] as const;

export function consistencyLabelKey(pct: number | null): string {
  if (pct === null) return 'rhythm.consistency.noTarget';
  if (pct >= 80) return 'rhythm.consistency.steady';
  if (pct >= 55) return 'rhythm.consistency.rhythmic';
  if (pct >= 30) return 'rhythm.consistency.varies';
  return 'rhythm.consistency.thin';
}

export interface CompositionSegment {
  key: CompositionKey;
  pct: number;
}

/** Builds the macro composition (kcal-share) segments for the calorie pill. */
export function buildComposition(macros: MacroPattern[]): {
  totalKcal: number;
  segments: CompositionSegment[];
} {
  const composition = COMPOSITION_KEYS.map((key) => {
    const macro = macros.find((m) => m.key === key);
    if (!macro || macro.averagePerDay <= 0) {
      return { key, kcal: 0 };
    }
    return { key, kcal: macro.averagePerDay * KCAL_PER_GRAM[key] };
  });
  const totalKcal = composition.reduce((sum, c) => sum + c.kcal, 0);
  const segments = composition.map((c) => ({
    key: c.key,
    pct: totalKcal > 0 ? (c.kcal / totalKcal) * 100 : 0,
  }));
  return { totalKcal, segments };
}

/** The macro rows shown under the calorie pill, in the web's fixed order. */
export function orderedMacroRows(macros: MacroPattern[]): MacroPattern[] {
  return ORDERED_MACROS.map((key) => macros.find((m) => m.key === key)).filter(
    (m): m is MacroPattern => Boolean(m)
  );
}

File: apps/mobile-flutter/lib/features/nutrition/logic/rhythm_logic.dart
/// Pure daily-rhythm helpers. The composition primitives now live in
/// `shared/logic/macro_composition.dart` and are re-exported here so existing
/// call sites keep reading; the web side is `COMPOSITION_COLORS` in
/// `components/nutrition/sections/macro-trend-utils.ts` (keep in sync).
///
/// The `daily-rhythm.tsx` section the rest of this was vendored from was
/// removed in the web nutrition rewrite (commit cdd7a3fa); parity of the
/// remaining helpers has not been re-verified since.
library;

import '../../../models/nutrition/nutrition.dart';
import '../../../shared/logic/macro_composition.dart';

export '../../../shared/logic/macro_composition.dart';


const List<String> kOrderedMacros = ['protein', 'carbohydrate', 'fat', 'fiber'];

String consistencyLabelKey(double? pct) {
  if (pct == null) return 'rhythm.consistency.noTarget';
  if (pct >= 80) return 'rhythm.consistency.steady';
  if (pct >= 55) return 'rhythm.consistency.rhythmic';
  if (pct >= 30) return 'rhythm.consistency.varies';
  return 'rhythm.consistency.thin';
}

/// Builds the macro composition (kcal-share) segments for the calorie pill.
Composition buildComposition(List<MacroPattern> macros) {
  final composition = kCompositionKeys.map((key) {
    final macro = macros.where((m) => m.key == key).firstOrNull;
    if (macro == null || macro.averagePerDay <= 0) {
      return (key: key, kcal: 0.0);
    }
    return (key: key, kcal: macro.averagePerDay * kKcalPerGram[key]!);
  }).toList();
  final totalKcal = composition.fold<double>(0, (sum, c) => sum + c.kcal);
  final segments = composition
      .map((c) => CompositionSegment(
            key: c.key,
            pct: totalKcal > 0 ? (c.kcal / totalKcal) * 100 : 0,
          ))
      .toList();
  return Composition(totalKcal: totalKcal, segments: segments);
}

/// The macro rows shown under the calorie pill, in the web's fixed order.
List<MacroPattern> orderedMacroRows(List<MacroPattern> macros) {
  return kOrderedMacros
      .map((key) => macros.where((m) => m.key == key).firstOrNull)
      .whereType<MacroPattern>()
      .toList();
}
